File: apps/reading/src/api/paidPurchaseHtmlViewSessionCompose.ts
/**
 * Paid purchase → HTML view session compose (pure).
 *
 * Operator vision: after marketplace paid (or free) gate is ready, open an
 * HTML-native reading session for the book in-account — never PDF primary.
 *
 * purchase_executed / charge_executed / hosted always false.
 * pdf_view_authorized always false.
 * store_mutated always false.
 */

import {
  composeMarketplacePaidPurchaseGate,
  type MarketplacePaidPurchaseGateCompose,
} from "./marketplacePaidPurchaseGateCompose";
import {
  composeHtmlAssetViewSession,
  type HtmlAssetViewSessionCompose,
} from "./htmlAssetViewSessionCompose";

export interface PaidPurchaseHtmlViewSessionInput {
  session_id: string;
  /** Asset id for the HTML view session (account-hosted book id). */
  asset_id: string;
  title: string;
  account_id: string;
  free_copy_available: boolean | null;
  free_html_projection_sha?: string | null;
  purchase_html_projection_sha?: string | null;
  port_requested: boolean;
  purchase_ack: boolean;
  list_price_usd: number | null;
  approved_spend_usd: number | null;
  remaining_budget_usd: number | null;
  operator_ack: boolean;
  view_requested: boolean;
  twin_bound?: boolean;
  twin_substrate_ready?: boolean;
  claimed_format?: string | null;
}

export interface PaidPurchaseHtmlViewSessionCompose {
  purchase_gate: MarketplacePaidPurchaseGateCompose;
  view: HtmlAssetViewSessionCompose | null;
  /**
   * True when purchase/free gate_ready and html_view session_ready.
   */
  session_package_ready: boolean;
  purchase_executed: false;
  charge_executed: false;
  hosted: false;
  pdf_view_authorized: false;
  store_mutated: false;
  notes: string[];
  authority: "paid_purchase_html_view_session_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose marketplace gate + HTML view session for seamless book open.
 * Never purchases, hosts, or authorizes PDF.
 */
export function composePaidPurchaseHtmlViewSession(
  input: PaidPurchaseHtmlViewSessionInput,
): PaidPurchaseHtmlViewSessionCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (typeof input.view_requested !== "boolean") {
    throw new Error("view_requested must be an explicit boolean");
  }

  const session_id = requireNonEmpty(input.session_id, "session_id");
  const asset_id = requireNonEmpty(input.asset_id, "asset_id");

  const notes: string[] = [
    "purchase_executed=false · charge_executed=false · hosted=false",
    "pdf_view_authorized=false — HTML-native only",
    "store_mutated=false — pure session pack",
  ];

  const purchase_gate = composeMarketplacePaidPurchaseGate({
    title: input.title,
    account_id: input.account_id,
    free_copy_available: input.free_copy_available,
    free_html_projection_sha: input.free_html_projection_sha,
    purchase_html_projection_sha: input.purchase_html_projection_sha,
    port_requested: input.port_requested,
    purchase_ack: input.purchase_ack,
    list_price_usd: input.list_price_usd,
    approved_spend_usd: input.approved_spend_usd,
    remaining_budget_usd: input.remaining_budget_usd,
    operator_ack: input.operator_ack,
  });
  notes.push(...purchase_gate.notes);

  // Prefer sha from free_port when gate ready
  const sha =
    purchase_gate.free_port.html_projection_sha ??
    input.purchase_html_projection_sha ??
    input.free_html_projection_sha ??
    null;

  let view: HtmlAssetViewSessionCompose | null = null;
  if (purchase_gate.gate_ready || input.view_requested) {
    view = composeHtmlAssetViewSession({
      session_id,
      asset_id,
      html_projection_sha: sha,
      view_requested: input.view_requested,
      twin_bound: input.twin_bound === true,
      twin_substrate_ready: input.twin_substrate_ready,
      claimed_format: input.claimed_format,
    });
    notes.push(...view.notes);
  } else {
    notes.push("view session not composed — gate not ready and view not requested");
  }

  const session_package_ready =
    purchase_gate.gate_ready &&
    view != null &&
    view.session_ready === true &&
    view.pdf_view_authorized === false;

  if (!purchase_gate.gate_ready) {
    notes.push("session_package_ready=false — marketplace gate not ready");
  } else if (view == null || !view.session_ready) {
    notes.push("session_package_ready=false — HTML view session not ready");
  } else {
    notes.push(
      "session_package_ready=true — HTML open intent; still not purchased/hosted",
    );
  }

  if (
    purchase_gate.purchase_executed !== false ||
    purchase_gate.charge_executed !== false ||
    purchase_gate.hosted !== false ||
    purchase_gate.pdf_view_authorized !== false
  ) {
    throw new Error("invariant: purchase gate honesty flags must remain false");
  }
  if (
    view != null &&
    (view.pdf_view_authorized !== false || view.store_mutated !== false)
  ) {
    throw new Error("invariant: view honesty flags must remain false");
  }

  notes.push("purchase_executed=false");
  notes.push("charge_executed=false");
  notes.push("hosted=false");
  notes.push("pdf_view_authorized=false");
  notes.push("store_mutated=false");

  return {
    purchase_gate,
    view,
    session_package_ready,
    purchase_executed: false,
    charge_executed: false,
    hosted: false,
    pdf_view_authorized: false,
    store_mutated: false,
    notes,
    authority: "paid_purchase_html_view_session_compose_advisory",
  };
}

export function formatPaidPurchaseHtmlViewSessionSummary(
  c: PaidPurchaseHtmlViewSessionCompose,
): string {
  return (
    `session_package_ready=${c.session_package_ready} · ` +
    `gate_ready=${c.purchase_gate.gate_ready} · ` +
    `html_view_ready=${c.view ? c.view.html_view_ready : "n/a"} · ` +
    `purchase_executed=false · charge_executed=false · hosted=false · ` +
    `pdf_view_authorized=false · store_mutated=false`
  );
}

File: substrate/paid_purchase_html_view_session_compose.py
"""Paid purchase → HTML view session compose (pure).

purchase_executed, charge_executed, hosted, pdf_view_authorized, store_mutated
always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.html_asset_view_session_compose import (
    HtmlAssetViewSessionCompose,
    HtmlAssetViewSessionComposeError,
    compose_html_asset_view_session,
)
from substrate.marketplace_paid_purchase_gate_compose import (
    MarketplacePaidPurchaseGateCompose,
    MarketplacePaidPurchaseGateComposeError,
    compose_marketplace_paid_purchase_gate,
)


class PaidPurchaseHtmlViewSessionComposeError(ValueError):
    """Fail-closed validation for paid purchase HTML view session."""


@dataclass(frozen=True)
class PaidPurchaseHtmlViewSessionCompose:
    purchase_gate: MarketplacePaidPurchaseGateCompose
    view: HtmlAssetViewSessionCompose | None
    session_package_ready: bool
    purchase_executed: bool
    charge_executed: bool
    hosted: bool
    pdf_view_authorized: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "purchase_gate": self.purchase_gate.to_dict(),
            "view": self.view.to_dict() if self.view is not None else None,
            "session_package_ready": self.session_package_ready,
            "purchase_executed": False,
            "charge_executed": False,
            "hosted": False,
            "pdf_view_authorized": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "paid_purchase_html_view_session_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise PaidPurchaseHtmlViewSessionComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_paid_purchase_html_view_session(
    *,
    session_id: object,
    asset_id: object,
    title: object,
    account_id: object,
    free_copy_available: object,
    port_requested: object,
    purchase_ack: object,
    list_price_usd: object,
    approved_spend_usd: object,
    remaining_budget_usd: object,
    operator_ack: object,
    view_requested: object,
    free_html_projection_sha: object | None = None,
    purchase_html_projection_sha: object | None = None,
    twin_bound: object | None = None,
    twin_substrate_ready: object | None = None,
    claimed_format: object | None = None,
) -> PaidPurchaseHtmlViewSessionCompose:
    """Compose marketplace gate + HTML view session. Never charges/hosts/PDF."""
    if not isinstance(operator_ack, bool):
        raise PaidPurchaseHtmlViewSessionComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(view_requested, bool):
        raise PaidPurchaseHtmlViewSessionComposeError(
            "view_requested must be an explicit boolean"
        )

    session = _require_nonempty(session_id, field="session_id")
    asset = _require_nonempty(asset_id, field="asset_id")

    notes: list[str] = [
        "purchase_executed=false · charge_executed=false · hosted=false",
        "pdf_view_authorized=false — HTML-native only",
        "store_mutated=false — pure session pack",
    ]

    try:
        purchase_gate = compose_marketplace_paid_purchase_gate(
            title=title,
            account_id=account_id,
            free_copy_available=free_copy_available,
            free_html_projection_sha=free_html_projection_sha,
            purchase_html_projection_sha=purchase_html_projection_sha,
            port_requested=port_requested,
            purchase_ack=purchase_ack,
            list_price_usd=list_price_usd,
            approved_spend_usd=approved_spend_usd,
            remaining_budget_usd=remaining_budget_usd,
            operator_ack=operator_ack,
        )
    except MarketplacePaidPurchaseGateComposeError as e:
        raise PaidPurchaseHtmlViewSessionComposeError(str(e)) from e
    notes.extend(purchase_gate.notes)

    sha = purchase_gate.free_port.html_projection_sha
    if sha is None:
        if purchase_html_projection_sha is not None:
            sha = purchase_html_projection_sha  # type: ignore[assignment]
        elif free_html_projection_sha is not None:
            sha = free_html_projection_sha  # type: ignore[assignment]

    view: HtmlAssetViewSessionCompose | None = None
    if purchase_gate.gate_ready or view_requested:
        try:
            view = compose_html_asset_view_session(
                session_id=session,
                asset_id=asset,
                html_projection_sha=sha,
                view_requested=view_requested,
                twin_bound=True if twin_bound is True else False,
                twin_substrate_ready=twin_substrate_ready,
                claimed_format=claimed_format,
            )
        except HtmlAssetViewSessionComposeError as e:
            raise PaidPurchaseHtmlViewSessionComposeError(str(e)) from e
        notes.extend(view.notes)
    else:
        notes.append(
            "view session not composed — gate not ready and view not requested"
        )

    session_package_ready = (
        purchase_gate.gate_ready
        and view is not None
        and view.session_ready is True
        and view.pdf_view_authorized is False
    )
    if not purchase_gate.gate_ready:
        notes.append("session_package_ready=false — marketplace gate not ready")
    elif view is None or not view.session_ready:
        notes.append(
            "session_package_ready=false — HTML view session not ready"
        )
    else:
        notes.append(
            "session_package_ready=true — HTML open intent; still not purchased/hosted"
        )

    if (
        purchase_gate.purchase_executed is not False
        or purchase_gate.charge_executed is not False
        or purchase_gate.hosted is not False
        or purchase_gate.pdf_view_authorized is not False
    ):
        raise PaidPurchaseHtmlViewSessionComposeError(
            "invariant: purchase gate honesty flags must remain false"
        )
    if view is not None and (
        view.pdf_view_authorized is not False or view.store_mutated is not False
    ):
        raise PaidPurchaseHtmlViewSessionComposeError(
            "invariant: view honesty flags must remain false"
        )

    notes.extend(
        (
            "purchase_executed=false",
            "charge_executed=false",
            "hosted=false",
            "pdf_view_authorized=false",
            "store_mutated=false",
        )
    )

    return PaidPurchaseHtmlViewSessionCompose(
        purchase_gate=purchase_gate,
        view=view,
        session_package_ready=session_package_ready,
        purchase_executed=False,
        charge_executed=False,
        hosted=False,
        pdf_view_authorized=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="paid_purchase_html_view_session_compose_advisory",
    )


def format_paid_purchase_html_view_session_summary(
    c: PaidPurchaseHtmlViewSessionCompose,
) -> str:
    html_ready = c.view.html_view_ready if c.view is not None else "n/a"
    return (
        f"session_package_ready={c.session_package_ready} · "
        f"gate_ready={c.purchase_gate.gate_ready} · "
        f"html_view_ready={html_ready} · "
        f"purchase_executed=false · charge_executed=false · hosted=false · "
        f"pdf_view_authorized=false · store_mutated=false"
    )


__all__ = [
    "PaidPurchaseHtmlViewSessionCompose",
    "PaidPurchaseHtmlViewSessionComposeError",
    "compose_paid_purchase_html_view_session",
    "format_paid_purchase_html_view_session_summary",
]
